## Quality Control
This subtask demonstrates the quality control (QC) steps for analyzing raw sequencing data in FASTQ format. The dataset consists of Next-Generation Sequencing (NGS) samples, including one sample from normal tissue and one from cancer tissue. The primary goal is to evaluate the quality of these sequencing reads, which is crucial for downstream analysis such as alignment and variant calling.

## Data Overview

File Format: *FASTQ*

Samples:
1. `PA221MH-lib09-P19-Tumor_S2_L001_R1_001.fastq.gz`
2. `PA221MH-lib09-P19-Tumor_S2_L001_R2_001.fastq.gz`
3. `PA221MH-lib09-P19-Norm_S1_L001_R1_001.fastq.gz`
4. `PA221MH-lib09-P19-Norm_S1_L001_R2_001.fastq.gz`

Tool Used: `FastQC` and `MultiQC`

## Prerequisites
Before proceeding with the quality control, ensure you have the following tools installed:

***FastQC***: A tool for checking the quality of raw sequencing data.

***Browser***: To view FastQC HTML reports

## Running FastQC on NGS Data
Bioinformatic analysis begins with quality control of sequence data to ensure its reliability and accuracy. This step helps identify potential errors, contamination, or issues that may have occurred during sequencing or the experiment. Using tools like FastQC, the sequencing files are analyzed to assess their quality and ensure they meet the required standards for downstream analysis.

## Command to Run FastQC for the Samples
``` bash
mkdir fastqc_reports  # Create a directory for the reports
fastqc
PA220KH-lib09-P19-Tumor_S2_L001_R1_001.fastq.gz
PA220KH-lib09-P19-Tumor_S2_L001_R2_001.fastq.gz
PA221MH-lib09-P19-Norm_S1_L001_R2_001.fastq.gz
PA221MH-lib09-P19-Norm_S1_L001_R1_001.fastq.gz
-o Fastqc_results/
```
## Expected Output Files

For each FASTQ file, FastQC will generate:

`HTML Report`: A human-readable report summarizing the quality metrics. `here `_.

`.zip File`: A compressed archive containing detailed analysis data.


## HTML reports
HTML reports contains following modules:
1. Basic Statistics
2. Per base sequence quality
3. Per tile sequence quality
4. Per sequence quality scores
5. Per base sequence content
6. Per sequence GC content
7. Per base N content
8. Sequence Length Distribution
9. Sequence Duplication Levels
10. Overrepresented sequences
11. Adapter Content


